### PR TITLE
fix: display error when project is not found

### PR DESCRIPTION
**Pull request by agg23**
Display an error message when a project is not found, such as when its name changes at runtime.

### CODE REVIEW

File: packages/playwright/src/runner/dispatcher.ts [hunk 1]
```diff
@@ -412,11 +412,18 @@ class JobDispatcher {
...
     // - no unrecoverable worker error
-    if (!this._remainingByTestId.size && !this._failedTests.size && !params.fatalErrors.length && !params.skipTestsDueToSetupFailure.length && !params.fatalUnknownTestIds && !params.unexpectedExitError) {
+    if (!this._remainingByTestId.size && !this._failedTests.size && !params.fatalErrors.length && !params.skipTestsDueToSetupFailure.length && !params.fatalUnknownTestIds && !params.missingProjectById && !params.unexpectedExitError) {
       this._finished({ didFail: false });
       return;
     }
 
+    if (params.missingProjectById) {
+      this._failureTracker.onWorkerError();
+      this._reporter.onError?.({ message: `Project with name ${params.missingProjectById} not found. Make sure project name does not change.` });
```
Comment (dgozman) [inline packages/playwright/src/runner/dispatcher.ts:422]: - This place should already know the project through `this._job.projectId`.
- We should display the project name to the user, not the project id.

File: packages/playwright/src/worker/workerMain.ts [hunk 2]
```diff
@@ -103,13 +105,18 @@ export class WorkerMain extends ProcessRunner {
   }
 
   override async gracefullyClose() {
+    if (this._isMissingProject) {
```
Comment (dgozman) [inline packages/playwright/src/worker/workerMain.ts:108]: Let's instead move this to the line 115 (after calling `this._stop()`), and perhaps check `!this._config` which is equivalent to not have loaded once.

File: packages/playwright/src/worker/workerMain.ts [hunk 3]
```diff
@@ -186,28 +193,43 @@ export class WorkerMain extends ProcessRunner {
...
   async runTestGroup(runPayload: RunPayload) {
     this._runFinished = new ManualPromise<void>();
     const entries = new Map(runPayload.entries.map(e => [e.testId, e]));
-    let fatalUnknownTestIds;
+    let fatalUnknownTestIds: string[] | undefined;
     try {
-      await this._loadIfNeeded();
-      const fileSuite = await loadTestFile(runPayload.file, this._config.config.rootDir);
-      const suite = bindFileSuiteToProject(this._project, fileSuite);
+      const workerState = await this._load();
```
Comment (dgozman) [inline packages/playwright/src/worker/workerMain.ts:219]: I do not understand why do we have to use the return value now, instead of reading from `this._config` and similar. Perhaps just return a boolean, or better yet set `this._isMissingProject` inside `_loadIfNeeded()` and check it here?

File: packages/playwright/src/common/ipc.ts [hunk 4]
```diff
@@ -125,6 +125,7 @@ export type DonePayload = {
   fatalErrors: TestInfoErrorImpl[];
   skipTestsDueToSetupFailure: string[];  // test ids
   fatalUnknownTestIds?: string[];
+  missingProjectById?: string;
```
Comment (dgozman) [inline packages/playwright/src/common/ipc.ts:128]: I'd call this `fatalUnknownProjectId` to be consistent, and make it a boolean instead.

File: packages/playwright/src/worker/workerMain.ts [hunk 5]
```diff
@@ -186,28 +193,43 @@ export class WorkerMain extends ProcessRunner {
...
+  } | undefined> {
     if (this._config)
-      return;
+      return { config: this._config, project: this._project, poolBuilder: this._poolBuilder };
 
     this._config = await deserializeConfig(this._params.config);
-    this._project = this._config.projects.find(p => p.id === this._params.projectId)!;
+    const project = this._config.projects.find(p => p.id === this._params.projectId);
+    if (!project)
+      return undefined;
```
Comment (dgozman) [inline packages/playwright/src/worker/workerMain.ts:207]: After reading through this PR, I realized the following. If we just throw a nicely formatted error here, it will end up as a `fatalError` and will be presented to the user in a nice way and cancel any tests in this run. The only problem is we'll probably have multiple of them - one per test group, but that's not a big deal for such an edge case. WDYT?

File: packages/playwright/src/worker/workerMain.ts (hunk 3 above)
Comment (agg23) [reply packages/playwright/src/worker/workerMain.ts:219]: I considered that to be bad control flow, since clearly you could continue executing without checking `this._isMissingProject`. I'll change it though.

File: packages/playwright/src/worker/workerMain.ts [hunk 6]
```diff
@@ -191,14 +197,19 @@ export class WorkerMain extends ProcessRunner {
       return;
 
     this._config = await deserializeConfig(this._params.config);
-    this._project = this._config.projects.find(p => p.id === this._params.projectId)!;
+    const project = this._config.projects.find(p => p.id === this._params.projectId);
+    if (!project) {
+      this._isMissingProject = true;
```
Comment (dgozman) [inline packages/playwright/src/worker/workerMain.ts:202]: If you assign `this._config` after this check, we can get rid of `_isMissingProject` and replace it with `if (!!this._config)`. This would mean "if we never loaded config, there is nothing to cleanup anyway".

File: packages/playwright/src/worker/workerMain.ts [hunk 7]
```diff
@@ -191,14 +197,19 @@ export class WorkerMain extends ProcessRunner {
...
+      throw new Error(`Project with name "${this._params.projectId}" not found. Make sure project name does not change`);
```
Comment (dgozman) [inline packages/playwright/src/worker/workerMain.ts:203]:
```suggestion
      throw new Error(`Project "${this._params.projectId}" not found in the worker process. Make sure project name does not change.`);
```

**Review by dgozman — COMMENTED** [review]
Let's add a test for this!

**Review by dgozman — APPROVED** [review]
